sequencing-server/src/backgroundTranspiler.ts

```diff
@@ -0,0 +1,144 @@
+import { piscina, graphqlClient, typeCheckingCache, promiseThrottler } from './app.js';
+import { objectCacheKeyFunction } from './lib/batchLoaders/index.js';
+import crypto from 'crypto';
+import { generateTypescriptForGraphQLActivitySchema } from './lib/codegen/ActivityTypescriptCodegen.js';
+import DataLoader from 'dataloader';
+import { activitySchemaBatchLoader } from './lib/batchLoaders/activitySchemaBatchLoader.js';
+import { commandDictionaryTypescriptBatchLoader } from './lib/batchLoaders/commandDictionaryTypescriptBatchLoader.js';
+import type { typecheckExpansion } from './worker';
+import { Result } from '@nasa-jpl/aerie-ts-user-code-runner/build/utils/monads.js';
+import { getLatestCommandDictionary, getLatestMissionModel, getExpansionRule } from './utils/hasura.js';
+
+export async function backgroundTranspiler(numberOfThreads: number = 2) {
+  if (graphqlClient === null) {
+    return;
+  }
+
+  // Fetch latest mission model
+  const { mission_model_aggregate } = await getLatestMissionModel(graphqlClient);
+  if (!mission_model_aggregate) {
+    console.log(
+      '[ Background Transpiler ] Unable to fetch the latest mission model. Aborting background transpiling...',
+    );
+    return;
+  }
+
+  // Fetch latest command dictionary
+  const { command_dictionary_aggregate } = await getLatestCommandDictionary(graphqlClient);
+  if (!command_dictionary_aggregate) {
+    console.log(
+      '[ Background Transpiler ] Unable to fetch the latest command dictionary. Aborting background transpiling...',
+    );
+    return;
+  }
+
+  const commandDictionaryId = command_dictionary_aggregate.aggregate.max.id;
+  const missionModelId = mission_model_aggregate.aggregate.max.id;
+  const { expansion_rule } = await getExpansionRule(graphqlClient, missionModelId, commandDictionaryId);
+
+  if (expansion_rule === null || expansion_rule.length === 0) {
+    console.log(`[ Background Transpiler ] No expansion rules to transpile.
+    Aborting transpiling...`);
+    return;
+  }
+
+  const commandTypescriptDataLoader = new DataLoader(commandDictionaryTypescriptBatchLoader({ graphqlClient }), {
+    cacheKeyFn: objectCacheKeyFunction,
+    name: null,
+  });
+  const activitySchemaDataLoader = new DataLoader(activitySchemaBatchLoader({ graphqlClient }), {
+    cacheKeyFn: objectCacheKeyFunction,
+    name: null,
+  });
+
+  const commandTypes = await commandTypescriptDataLoader.load({
+    dictionaryId: missionModelId,
+  });
+
+  if (commandTypes === null) {
+    console.log(`[ Background Transpiler ] Unable to fetch command ts lib.
+    Aborting transpiling...`);
+    return;
+  }
+
+  // only process 'numberOfThreads' worth at a time ex. transpile 2 logics at a time
+  // This allows for expansion set and sequence expansion to utilize the remaining workers
+  for (let i = 0; i < expansion_rule.length; i += numberOfThreads) {
```

Comment:
Batching like this shouldn't be necessary with the promise throttler as it will simply queue up runs to execute max numberOfTreads at a time. So you should be able to just queue them all up without waiting on batches.

Reply:
Are you saying that the promise was getting inserted into the cache prior to resolving, and then being read out prior to resolving elsewhere?

I worry that this is a legitimate race condition that needs to be addressed and it's unclear to me how batching would address that.

Reply:
There are two parts where transpilation can occur: the background transpiler and expansion set creation. During expansion set creation, the promise generated by the PromiseThrottler is either cached or retrieved from the cache. If the promise isn't yet resolved, Promise.allSettled() guarantees that the PromiseThrottler processes and completes all cached promises.

expansion-set code
```ts
 const typecheckErrorPromises = await Promise.allSettled(
    expansions.map(async (expansion, index) => {
    
      const hash = ...
     
      if (typeCheckingCache.has(hash)) {
        return typeCheckingCache.get(hash);
      }

      const typeCheckResult = promiseThrottler.run(() => {
        return (
          piscina.run(
           ...
          ) as ReturnType<typeof typecheckExpansion>
        ).then(Result.fromJSON);
      });

      typeCheckingCache.set(hash, typeCheckResult);
      return typeCheckResult;
    }),
  );
```

For the above example when implementing batching. We have 120 expansion logic files to transpile, only two are fetched at a time. These logic are either cached or retrieved from the cache. The process then waits for both promises to finish before moving on to the next two. This approach avoids placing all 120 promises in the cache simultaneously and sequentially adds them 2 at a time. 


When creating an expansion set with logic files 118-120, the background transpiler has finished/cached  1-58 promises, and the next two 59-60 are pending. We will add 118-120 to the cache and promiseThrottler has 6 threads free to start processing 118-120. The only blocking `Promise.allSettled() ` is doing is for the 118-120 promises.

Reply:
On the other hand, creating an expansion set with all 120 logic files would queue all promises at once and wait for their completion. Although, the background transpiler can still assist by picking up pending promises from the shared cache and processing them, improving the overall transpilation time.

Reply:
I feel like there's a solution where we don't have to wait for all of the in process transpilations, but just the ones we need to use. Unfortunately, the promise throttler doesn't consider priority in it's current form, so we can't really "jump" an expansion that has been requested to the front of the queue.

An alternative solution only has the background transpiler pushing into the cache on completion/promise resolution, rather than pushing the unresolved promise (that could wait a long time for 100 other expansions to transpile). Thus the cache has either those expansions immediately requested as unresolved, or those that are background transpiled pushed in already resolved.

It's not a fully fleshed out solution, but perhaps it sparks some ideas for you @goetzrrGit.

Reply:
These are two-way doors where we can do better in a future change, so I don't want our pursuit of an optimal solution to delay a sufficient solution for Clipper.

Reply:
It does, I was thinking about this more last night. I considered changing the Throttler into a priority queue, where all promises generated by the backgroundTranspiler would be treated as lower priority than requests from the `expansion-set` or `expansion-to-sequences`. Additionally, I agree with your suggestion of only adding the completed promise to the cache instead of pending promises. I believe there is room for improvement in this initial implementation. 

Yeah, we are in a situation now where Clipper needs solution now so we have to punt more design and ideals down the line atm.